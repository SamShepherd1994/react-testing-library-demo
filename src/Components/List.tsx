import * as React from 'react';

type ListProps = {
  items: { id: string; name: string }[];
};

export const List: React.FC<ListProps> = ({ items }: ListProps) => {
  if (items.length === 0) {
    return <div>No Items Found.</div>;
  }

  return (
    <table>
      <tbody>
        {items.map((item) => (
          <tr key={item.id.toString()}>
            <td>{item.name}</td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};
