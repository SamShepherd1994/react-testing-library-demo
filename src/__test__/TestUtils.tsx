import { act, render, RenderOptions } from '@testing-library/react';
import {
  Renderer,
  renderHook,
  RenderHookOptions,
  RenderHookResult,
} from '@testing-library/react-hooks';
import { ReactElement } from 'react';
import { SearchContext } from '../Contexts/DefaultSearchContext';

const Providers: React.FC = ({ children }) => {
  return (
    <SearchContext.Provider
      value={{ defaultValue: 'this is not the usual default value...' }}
    >
      {children}
    </SearchContext.Provider>
  );
};

export const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: Providers, ...options });

/**
 * Can be used when rendering a hook results in an error stating that state updates must be performed in an act.
 * E.g. If a hook updates some state on initial render
 * @param hook
 * @param options
 * @returns
 */
export const renderHookInAct = async <T, R>(
  hook: (props: T) => R,
  options?: RenderHookOptions<T>
) => {
  let result: RenderHookResult<T, R, Renderer<T>> | undefined;
  await act(
    async () =>
      await new Promise((resolve) => {
        const rendered = renderHook(hook, options);
        result = rendered;
        // small timeout is used to ensure state has been updated suggested in https://github.com/testing-library/react-hooks-testing-library/issues/656
        setTimeout(resolve, 10);
      })
  );
  return result;
};
