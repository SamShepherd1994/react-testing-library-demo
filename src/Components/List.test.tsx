import { customRender } from '../__test__/TestUtils';
import { List } from './List';

describe('List', () => {
  it('Renders No items found when list is empty', () => {
    const { asFragment } = customRender(<List items={[]} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with single item', () => {
    const { asFragment } = customRender(
      <List items={[{ id: '1', name: 'name' }]} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with multiple items', () => {
    const { asFragment } = customRender(
      <List
        items={[
          { id: '1', name: 'name' },
          { id: '2', name: 'name2' },
        ]}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
