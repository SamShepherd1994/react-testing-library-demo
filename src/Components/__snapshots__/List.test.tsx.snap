// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Renders No items found when list is empty 1`] = `
<DocumentFragment>
  <div>
    No Items Found.
  </div>
</DocumentFragment>
`;

exports[`List renders correctly with multiple items 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td>
          name
        </td>
      </tr>
      <tr>
        <td>
          name2
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`List renders correctly with single item 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td>
          name
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
